core/state/snapshot/generate.go

```diff
@@ -159,169 +181,409 @@ func journalProgress(db ethdb.KeyValueWriter, marker []byte, stats *generatorSta
 	rawdb.WriteSnapshotGenerator(db, blob)
 }
 
+// proofResult contains the output of range proving which can be used
+// for further processing no matter it's successful or not.
+type proofResult struct {
+	keys     [][]byte // The key set of all elements being iterated, even proving is failed
+	vals     [][]byte // The val set of all elements being iterated, even proving is failed
+	cont     bool     // Indicator if there exists more elements in the range, only meaningful when proving is successful
+	proofErr error    // Indicator whether the given state range is valid or not
+}
+
+// valid returns the indicator that range proof is successful or not.
+func (result *proofResult) valid() bool {
+	return result.proofErr == nil
+}
+
+// last returns the last verified element key no matter the range proof is
+// successful or not. Nil is returned if nothing involved in the proving.
+func (result *proofResult) last() []byte {
+	var last []byte
+	if len(result.keys) > 0 {
+		last = result.keys[len(result.keys)-1]
+	}
+	return last
+}
+
+// forEach iterates all the visited elements and applies the given callback on them.
+// The iteration is aborted if the callback returns non-nil error.
+func (result *proofResult) forEach(callback func(key []byte, val []byte) error) error {
+	if callback == nil {
+		return nil
+	}
+	for i := 0; i < len(result.keys); i++ {
+		key, val := result.keys[i], result.vals[i]
+		if err := callback(key, val); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// proveRange proves the state segment with particular prefix is "valid".
+// The iteration start point will be assigned if the iterator is restored from
+// the last interruption. Max will be assigned in order to limit the maximum
+// amount of data involved in each iteration.
+//
+// The proof result will be returned if the range proving is finished, otherwise
+// the error will be returned to abort the entire procedure.
+func (dl *diskLayer) proveRange(root common.Hash, tr *trie.Trie, prefix []byte, kind string, origin []byte, max int, valueConvertFn func([]byte) ([]byte, error)) (*proofResult, error) {
+	var (
+		keys  [][]byte
+		vals  [][]byte
+		proof = rawdb.NewMemoryDatabase()
+	)
+	iter := dl.diskdb.NewIterator(prefix, origin)
+	defer iter.Release()
+
+	for iter.Next() && len(keys) < max {
+		key := iter.Key()
+		if len(key) != len(prefix)+common.HashLength {
+			continue
+		}
+		keys = append(keys, common.CopyBytes(key[len(prefix):]))
+
+		if valueConvertFn == nil {
+			vals = append(vals, common.CopyBytes(iter.Value()))
+			continue
+		}
+		val, err := valueConvertFn(iter.Value())
+		if err != nil {
+			log.Debug("Failed to convert the flat state", "kind", kind, "key", common.BytesToHash(key[len(prefix):]), "error", err)
+			return nil, err
+		}
+		vals = append(vals, val)
+	}
+	// The snap state is exhausted, pass the entire key/val set for verification
+	if origin == nil && len(keys) == max {
```

Comment:
The `len(keys)==max` is wrong, should be `<`. I'll fix in a follow-up commit

Reply:
There is a more subtle flaw here aswell: checking `len(keys)==max` is not a sufficient condition, because the snap might hold exactly `max`items. So it's exhausted _and_ filled up the max. So I'm adding a more elaborate exit condition